### DISCUSSION TITLE
How exactly does the restart capability work?

### DISCUSSION THREAD

**Discussion by bobsalko** [snl-dakota/dakota#117] [Q&A]
I am running a very costly sensitivity study involving tens of thousands of simulations.  I will need to use the restart capability due to system policy limitations.  To give a specific example, the restart file has ~24,000 evaluations in it.  I would like to restart from there and run only one more evaluation and generate statistics to see where I am at.  Then run it for another 10k evaluations to see how the statistics change.  To do this, do I just change my number of samples to 1 and then run it as:

```dakota -i dakota.in -o dakota.out -read_restart dakota.rst -write_restart new.rst```

It first reads in all 24,000 previous evaluations from the restart file, then it runs 38 more cases (because I have 36 variables and (36+2)*1=38 as required for generating Sobol indices).  Then it generates statistics.  The output file says "38 new, 0 duplicate".  I just want to make sure those statistics are actually considering the other 24,000 evaluations from the restart file and not just the one new sample.  

I am a little confused about what I read in the user manual because it says Dakota will start from the beginning and use duplicate detection to catch duplicates.  Does this mean I should set number of samples to 632 (which would lead to 632*(36+2)=24,016 evaluations) and expect Dakota to just skip the first 631 samples which already exist in the restart file?  It actually ran the 38 cases when I tried to restart it with only 1 sample rather than skipping them because they are duplicates, so I'm not sure that is the way it works.

**Comment by briadam** [comment]
For the case where you've made no changes to the input file, the way basic restart should work is: Suppose your study specifies total of 100 samples, 80 completed before the run terminated. Now you come back and restart. Dakota will read 80 evaluations from the restart file. It will then start the sampling method from the beginning and as the method request samples, it will first check the imported restart data. For the 80 runs that completed (which might not correspond to contiguous evaluation IDs), it will report that a duplicate is detected and not run your analysis driver. The other 20 will not be found in restart and get sent out for evaluation.

In your case, I'm guessing you're getting foiled by changing the number of samples. Assuming you're using the `sampling` method (sample type LHS or Monte Carlo), when you change the number of samples specified in the input file, you're generating an entirely new design of experiments (set of points in the parameter space). In the example above, if you change samples from 100 to 101, the variable values of all 101 samples will almost surely be different than the previous 100 and no matches will be found in the restart data. This is due to the sampling method respecting the Latin design structure and/or restricted pairing to match the prescribed correlation (or lack of correlation).

Instead of changing the number of samples, you might instead want to consider using [refinement samples](https://snl-dakota.github.io/docs/6.19.0/users/usingdakota/reference/method-sampling-refinement_samples.html). For LHS the samples must double, but for Monte Carlo, you can arbitrarily increase with this mechanism. I think using this you can achieve the use case you described.

**Reply by bobsalko** [reply]
Ok, I tested the restart capability for a simpler problem in `dakota/share/dakota/examples/official/global_sensitivity/direct_sampling/direct_sample.in` and it works like this.  I also tested using refinement_samples and that works when I shut off `variance_based_decomp`; however, it does not seem to work as documented when `variance_based_decomp` is on.  To be clear, I am running this input:

```
# Global Sensitivity via Direct Sampling

environment,
	tabular_data
		tabular_data_file = 'samples.dat'
        custom_annotated header


method
    output normal
	sampling
	  samples = 100
       #refinement_samples = 100
	  seed = 6347
	  sample_type lhs
      variance_based_decomp

variables,
	uniform_uncertain =  3
	lower_bounds    =    -3.14159 -3.14159 -3.14159
	upper_bounds    =    +3.14159 +3.14159 +3.14159
	descriptors     =    'x'      'y'      'z'


model single

interface
    fork
    analysis_driver = 'python3 Ishigami.py'
    asynchronous evaluation_concurrency = 5


responses
    response_functions = 1
    descriptors = 'Ishigami'
	no_gradients
	no_hessians
```

as `dakota -i direct_sample.in` the first time to generate 100 samples (note actually 500 evaluations are done).  Then I uncomment the `refinement_samples = 100` line and run it again as:

`dakota -i direct_sample.in -read_restart dakota.rst -write_restart lhs100.rst`

And it immediately finishes.  All 500 evaluations are detected as duplicates and the same statistics are generated.  No new evaluations are performed.  I also tried setting `refinement_samples=500` just in case it was expecting that to match the number of actual evaluations but that also did not work.

**Comment by briadam** [comment]
I apologize Bob, I should have thought through the VBD use case implied in your original post... Dakota has a limitation that incremental (refinement) sampling isn't supported in conjunction with sampling-based VBD. There's no technical barrier, just not implemented. I can't immediately think of a workaround that would work with sampling-based VBD.

One idea might be to try PCE-based VBD, where you conduct a sampling design of experiments and then have Dakota build a polynomial chaos expansion from the samples and then the Sobol indices are computed analytically from the PCE. I believe with this approach you could import and reuse the data you've already generated, as well as generate more new samples. You can instruct Dakota to automatically build the best PCE it can from the data without overfitting via a cross-validation approach.

I think the closest example in the repo is here: https://github.com/snl-dakota/dakota-examples/tree/master/official/global_sensitivity/pce_on_data. I'm not quite sure how you do this from restart (vs. converting the restart to tabular data, which isn't hard with `dakota_restart_util`, but seems an extra burden).

I'll likely have to defer to @jadamstephens and others on the team from here...

Brian
